source/ShopPanel.cpp

```diff
@@ -222,81 +222,212 @@ void ShopPanel::CheckForMissions(Mission::Location location)
+void ShopPanel::BuyToCargo()
 {
-	return (playerShip && selectedOutfit && player.Cargo().Get(selectedOutfit))
-		|| player.Storage().Get(selectedOutfit);
 }
 
 
 
-bool ShopPanel::ShouldHighlight(const Ship *ship)
+void ShopPanel::Buy()
 {
-	return (hoverButton == 's');
 }
 
 
 
-void ShopPanel::DrawKey()
+void ShopPanel::Sell(bool /* storeOutfits */)
 {
 }
 
 
 
-int ShopPanel::VisibilityCheckboxesSize() const
+void ShopPanel::Sell()
 {
-	return 0;
 }
 
 
 
-void ShopPanel::ToggleForSale()
+ShopPanel::TransactionResult ShopPanel::CanInstall() const
 {
-	CheckSelection();
-	delayedAutoScroll = true;
+	return false;
 }
 
 
 
-void ShopPanel::ToggleStorage()
+void ShopPanel::Install()
 {
-	CheckSelection();
-	delayedAutoScroll = true;
 }
 
 
 
-void ShopPanel::ToggleCargo()
+ShopPanel::TransactionResult ShopPanel::CanUninstall() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::Uninstall()
 {
-	CheckSelection();
-	delayedAutoScroll = true;
+}
+
+
+
+bool ShopPanel::CanMoveToCargo() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::MoveToCargo()
+{
+}
+
+
+
+ShopPanel::TransactionResult ShopPanel::CanMoveToStorage() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::MoveToStorage()
+{
+}
+
+
+
+bool ShopPanel::CanSellMultiple() const
+{
+	return true;
+}
+
+
+
+ShopPanel::TransactionResult ShopPanel::CanDoTransaction(const SDL_Keycode key) const
+{
+	switch(key)
+	{
+		case 'b': // _Buy
+			return CanBuy();
+		case 's': // _Sell
+			return CanSell();
+		case 'r': // Sell Ship and put Outfits in Storage / Move to Storage
+			if(isOutfitter)
+				return CanMoveToStorage();
+			// Shipyard, sell Ship, send Cargo to Storage
+			return CanSell();
+		default:
+			break;
+	}
+
+	if(isOutfitter)
+	{
+		switch(key)
+		{
+			case 'c': // Move to/Buy to _Cargo
+			{
+				const auto canMove = CanMoveToCargo();
+				const auto canBuy = CanBuyToCargo();
+				if(canMove || canBuy)
+					return true;
+				if(!canBuy)
+					return canBuy; // Return why we cannot buy
+			}
+			break;  // debug for now
+			case 'u': // _Uninstall
+				return CanUninstall();
+			case 'i': // _Install
+				return CanInstall();
+			default:
+				return false;
+		}
+	}
+
+	return false;
+}
+
+
+
+void ShopPanel::DoTransaction(SDL_Keycode key)
+{
+	int modifier = CanSellMultiple() ? Modifier() : 1;
+
+	switch(key)
+	{
+		case 's': // _Sell
+			if(isOutfitter)
+				for(int i = 0; i < modifier && CanSell(); ++i)
+						Sell();
+			else
+				// Sell ship and outfits
+				Sell(false);
+			break;
+		case 'r': // Ship and/or Move outfits to Storage
+			if(isOutfitter)
+				for(int i = 0; i < modifier && CanMoveToStorage(); ++i)
+					MoveToStorage();
+			else
+				// Sell ship, Retain outfits
+				Sell(true);
+			break;
+		case 'c': // to _Cargo
+			if(isOutfitter)
+			{
+				// Move to Cargo or else Buy to _Cargo
+				if(CanMoveToCargo())
+				{
+					for(int i = 0; i < modifier && CanMoveToCargo(); ++i)
+						MoveToCargo();
+				}
+				else
+					BuyToCargo(); // modifier applied within
```

Comment:
What does this comment mean?

Reply:
See latest commit, I move all the modifier logic out of ShopPanel into the Shipyard/Outfitter and revised all comments.